Measurement

ichristen opened the pull request.

ichristen committed:
Changes to prefs:
 + prefs can now command (read, write, addlistener) pref_handler via 
function handles (read_fn, writ_fn, listen_fn) that are set in bind(),
 + prefs now have readProp() and writProp() methods that  are slightly 
less expensive than pre() and post()
 + pref_handler now suggests fixing prop.set methods rather than 
demanding removal.
 + pref_handler.prop_listener_ctrl was optimized.
 + Other minor changes.

ichristen committed:
Other changes to prefs and subclasses.
 + Made Prefs.Numeric superclass to handle min/max and flag presence.
 + Started modifying MultipleChoice to accept numerical inputs (e.g. mc 
= 1 ==> first choice).
 + Some performance improvements (e.g. validateattributes is expensive 
in Double).

ichristen committed:
Added Base.measurement. See foreward comment for explaination. This is not hooked up to anything yet.

ichristen commented on +Base/GetClasses.m:
```diff
@@ -6,7 +6,6 @@
 if any(arrayfun(@(d)strcmpi(fullfile(d.folder,d.name),target),dir))
     files(1) = []; %assume first entry is the present working directory  and is thus redundant
 end
-assert(length(files)==1,sprintf('Did not find a single match for %s - found %i',target,length(files)))
```
Empty file lists were already handled later. This assert breaks things rather than returning info to the user.

mwalsh161 replied:
this was actually intended to prevent further errors with more than 1 file returned

ichristen replied:
Fixed to > 1 in recent commits.

ichristen commented on +Base/GetClasses.m:
```diff
@@ -36,11 +35,8 @@
         visible = ~Base.EndsWith(class_str,'_invisible');
-        try %#ok<TRYNC>
-            visible = and(visible,~logical(eval(sprintf('%s%s.invisible',prefix,class_str))));
-        end
```
Feature NotImplemented Re: Michael.

mwalsh161 replied:
do we want to implement or should we scrap? implementing would require grabbing the property meta class and seeing if it was defined in this class.

ichristen replied:
Scrap.

ichristen commented on +Base/pref.m:
```diff
@@ -52,25 +52,42 @@
+        writ_fn     = [];                   % Calls tf = pref_handler.writProp(prop, val), with the appropriate arguments already filled in.
+        
+        listen_fn   = [];                   % Calls pref_handler.addlistener(prop, event, callback), with the appropriate arguments already filled in.
+    end
```
New function handles to hook back into pref_handler

ichristen commented on +Base/pref.m:
```diff
@@ -52,25 +52,42 @@
+        listen_fn   = [];                   % Calls pref_handler.addlistener(prop, event, callback), with the appropriate arguments already filled in.
+    end
+    properties (Hidden, SetAccess = private)
+        parent_class = '';                  % Stores the name of the parent pref_handler subclass for reference.
```
Useful info.

ichristen commented on +Base/pref.m:
```diff
@@ -96,11 +113,35 @@
+                    obj.writ_fn =       @(val)(module_instance.writProp(props(i), val));
+                    obj.listen_fn =     @(event, callback)(module_instance.addlistener(obj.property_name, event, callback));
+                end
+            end
```
Find our prop and make the function handles.

ichristen commented on +Base/pref.m:
```diff
@@ -96,11 +113,35 @@
+                    if ~isempty(mmethod)
+                        obj.(avail_methods{j}) = fnstring;
+                    end
+                end
```
To automatically find module.set.prop methods which have been renamed to module.set_prop; this only occurs when the pref.set is empty.

ichristen commented on +Base/pref.m:
```diff
@@ -197,6 +240,24 @@ function set_ui_value(obj,val)
+        function tf = isnumeric(~)
+            tf = false;
+        end
     end
```
Pulls the label from the ui.

ichristen commented on +Base/pref.m:
```diff
@@ -227,34 +285,45 @@ function set_ui_value(obj,val)
+                if ~mp.Hidden && ~mp.Abstract
+                    obj.(mp.Name) = p.Results.(mp.Name);
+                end
             end
```
Changed to go entirely through metaclass rather than properties. Much faster now. Also changed to ignore validation from Hidden functions. Not sure if I've thought of all the repercussions of this, but it works fine.

ichristen commented on +Base/pref.m:
```diff
@@ -227,34 +285,45 @@ function set_ui_value(obj,val)
     end
     methods
-        function summary = validation_summary(obj,indent,varargin)
+        function summary = validationSummary(obj,indent,varargin)
```
Started to move to camelcase. Will eventually do more.

ichristen commented on +Base/pref.m:
```diff
@@ -329,20 +398,42 @@ function set_ui_value(obj,val)
-                if ~isempty(obj.custom_clean) && obj.initialized
-                    val = obj.custom_clean(val,obj);
-                end
             end
```
set -> validate -> clean ==> clean -> validate -> set -> validation

Re: previous discussion.

ichristen commented on +Base/pref.m:
```diff
@@ -329,20 +398,42 @@ function set_ui_value(obj,val)
+        end
+        function listener = addlistener(obj, event, callback)
+            listener = obj.listen_fn(event, callback);
+        end
```
Functions that grasp the handles.

ichristen commented on +Base/pref_handler.m:
```diff
@@ -58,25 +64,38 @@
+                        ['Cannot directly use set methods with class-based prefs! Please modify:', 10,...
+                         '    `function set.<property_name>(obj, val)` ==>', 10,...
+                         '    `function val = set_<property_name>(obj, val, ~)`', 10,...
+                         'and set `val = obj.<property_name>` at the end of the function.']);
```
Easy fix for .set methods

ichristen commented on +Base/pref_handler.m:
```diff
@@ -58,25 +64,38 @@
                     obj.set_meta_pref(prop.Name, pref);
+
+%                     pl.addPref(mc.Name, pref);
+                    pl.addPref(obj, pref);
```
Registers into the prefList

ichristen commented on +Base/pref_handler.m:
```diff
@@ -179,8 +199,41 @@ function set_meta_pref(obj,name,pref)
+
+    end
+    methods(Access = private)
+        function prop_listener_ctrl(obj,name,enabled)
```
Improve speed. Need to remove debug comments from this file. How does one do this on a pull request?

mwalsh161 replied:
you can directly submit more commits to this branch and they will be included

ichristen replied:
+1

ichristen commented on +Base/pref_handler.m:
```diff
@@ -271,5 +315,119 @@ function post(obj,prop,event)
+        end
+    end
+
+end
```
Just pre() post() tacked together.

ichristen commented on +Base/prefList.m:
```diff
@@ -0,0 +1,140 @@
+classdef prefList < handle
```
Need to document this file.

ichristen commented on +Prefs/+Inputs/LabelControlBasic.m:
```diff
@@ -35,6 +41,11 @@
+            if strcmp(obj.uistyle, 'checkbox') && strcmp(pref.units, '0/1')
+                labeltext = labeltext(1:end-6);
+            end
+            
```
Not sure if there's a better way to do this, but I don't want the default units to display beside checkboxes.

mwalsh161 replied:
why not?

ichristen replied:
"Label [0/1] [x]" <--checkbox is redundant because it's obviously 0/1 if there's a checkbox. This bothers me.

ichristen commented on +Prefs/+Inputs/LabelControlBasic.m:
```diff
@@ -18,10 +18,16 @@
+            
             if ~isempty(pref.units)
-                labeltext = sprintf('%s (%s)',pref.name,pref.units);
+                labeltext = sprintf('%s (%s)', name, pref.units);
```
This should be changed to square brackets.

mwalsh161 replied:
lol sure

ichristen commented on +Prefs/Double.m:
```diff
@@ -35,15 +36,18 @@ function set_ui_value(obj,val)
         end
         function validate(obj,val)
-            validateattributes(val,{'numeric'},{'scalar'})
+            if ~isnumeric(val) || ~isscalar(val)
```
Perfomance improvements.

ichristen commented on +Prefs/Empty.m:
```diff
@@ -0,0 +1,38 @@
+classdef Empty < Prefs.Numeric
```
Useful for backend.

mwalsh161 replied:
explain? why would an empty pref array not suffice?

ichristen replied:
I want it to return values to functions like units and labels. I want it to feel like a pref that the program to touch without actually having innards.

ichristen commented on +Prefs/Empty.m:
```diff
@@ -0,0 +1,38 @@
+    end
+
+    methods
+        function obj = Empty(varargin)
```
Needs documentation.

ichristen commented on +Prefs/Integer.m:
```diff
@@ -12,7 +12,7 @@
             validate@Prefs.Double(obj,val);
             % Only further check is integer (allow Inf/NaN as integer)
-            if ~isinf(val) || ~isnan(val)
+            if ~isinf(val) && ~isnan(val)
```
This was a bug.

ichristen commented on +Prefs/MultipleChoice.m:
```diff
@@ -71,6 +95,15 @@ function validate(obj,val)
+                    val = [];
+                elseif val <= length(obj.choices)
+                    val = obj.choices{val};
+                end
```
Maybe should error? I forgot what I was doing.

mwalsh161 commented on +Prefs/Double.m:
```diff
@@ -35,15 +36,18 @@ function set_ui_value(obj,val)
             end
-            assert(val <= obj.max, 'Cannot set value greater than max.')
-            assert(val >= obj.min, 'Cannot set value less than min.')
+            if val <= obj.max || val >= obj.min
```
why not just turn this into an error statement if you don't want to use the validate method?

ichristen replied:
assert evaluates the string whether or not the assertion is true or false. This was limiting performance. Now the strings are generated only if they need to be.

mwalsh161 commented on +Prefs/Numeric.m:
```diff
@@ -0,0 +1,17 @@
+classdef Numeric < Base.pref
```
what is this used for?

ichristen replied:
Basically, to tell other things that subclasses of this are 'scannable', i.e. I can put in numeric values, or poll min/max without things breaking.

ichristen committed:
Addresses comments in pull request. Imposes sane capitalization for Prefs.

ichristen committed:
More sane capitalization.

ebersin commented on +Base/measurement.m:
```diff
@@ -0,0 +1,786 @@
+%                             'meta', struct('notjunk', 'no duplicate'),   'many', 'a', 'junk', 'b', 'fields', 'c'),...
+%                     obj.blank()};
+                
+%             test = {struct('dat', 1, 'std', 2, 'meta', struct(), 'many', 'a', 'junk', 'b', 'fields', 'c')};
```
???

ebersin commented on +Base/measurement.m:
```diff
@@ -0,0 +1,786 @@
+                    %%%%% If this Measurement has an expected structure...
+                    
+%                     assert(~isfieldfast(raw_fields, 'dat'), 'dat may not be a field of d');
+%                     assert(~isfieldfast(raw_fields, 'std'), 'std may not be a field of d');
```
???

ichristen committed:
Finished rename of files. More commenting and cleanup.

ichristen committed:
Further cleanup.

ichristen committed:
Merge branch 'gitPanel' into sweeping

ichristen committed:
Base commit; n.b. renaming Scan -> Sweep.

ichristen committed:
Bugfix.

ichristen committed:
Real bugfix.

ichristen committed:
Started work on SweepEditer + ~finished migration Scan -> Sweep.

ichristen committed:
Bugfixes in PrefRegister from the migration prefList -> PrefRegister.

ichristen committed:
Merge branch 'gitPanel' into sweeping

ichristen committed:
Merge branch 'measurement' into sweeping

ichristen committed:
Commenting + clarity.

ichristen committed:
Merge branch 'measurement' into sweeping

ichristen committed:
Made PrefRegister properly singleton.

ichristen committed:
`SweepEditor` half finished; integrating `Measurement`:
 + `SweepEditor` core UI mostly present, but not all hooked up.
 + `Base.Numeric` prefs are now subclasses of `Base.Measurement`
 + Changed `pref.units` -> `pref.unit` to avoid conflict with 
`Measurement` (it's worth it).

ichristen committed:
Debugging `Measurement` hookup to `Prefs`. Trying to get `Measurement` hooked up to `Experiment`.

ichristen committed:
Merge branch 'gitPanel' into sweeping

ichristen committed:
Got Measurement working. Adding this to SweepEditor.
 + Renovated Spectrum.
 + Need to change path. Current behavior is annoying.
 + MultipleChoice is slightly broken.
 + Added some old code that I need to rework into CC.

ichristen committed:
Small changes to Editor

ichristen committed:
Mostly finished `MeasurementRegister`.

ichristen committed:
Merge branch 'sweeping' of https://github.com/mwalsh161/CommandCenter into sweeping

ichristen committed:
Continuing work on making the SweepEditor function as a system.

ichristen committed:
Merge branch 'dev' of https://github.com/mwalsh161/CommandCenter into sweeping

ichristen committed:
More fleshing out SweepEditor.

ichristen committed:
Merge branch 'dev' of https://github.com/mwalsh161/CommandCenter into sweeping

ichristen committed:
Merge branch 'dev' of https://github.com/mwalsh161/CommandCenter into sweeping

ichristen committed:
Last edits to SweepEditor before modernizing Sweep to mesh with Measurement.

ichristen committed:
Sweep modernized. Started SweepController. Other fleshing.

ichristen committed:
SweepViewer and SweepProcesssed mostly modernized and debugged.

ichristen committed:
Windows UI adjustments + debugging.

ichristen committed:
Lots of Sweep debugging.

ichristen committed:
Debugged listeners, pointers, etc

ichristen committed:
Debugging and polishing SweepViewer. Fixed normalization.

ichristen committed:
Removed unfinished features.

ichristen committed:
Merge branch 'sweeping' of https://github.com/mwalsh161/CommandCenter into sweeping

ichristen committed:
Make SweepEditor spawn ScanViewers.

ichristen committed:
Added more timing features.

ichristen committed:
Merge branch 'sweeping' of https://github.com/mwalsh161/CommandCenter into measurement

ichristen committed:
Refreshed Measurement branch with stuff from Sweeping. Considering new architecture.
